Lib/compression/zstd/__init__.py

```diff
@@ -0,0 +1,259 @@
+    ZstdDict.__init__ method.
+
+    *zstd_dict* is a ZstdDict object, the basis dictionary.
+    *samples* is an iterable of samples, a sample is a bytes-like object
+    representing a file.
+    *dict_size* is the dictionary's maximum size, in bytes.
+    *level* is the compression level expected to use in production. The
+    statistics for each compression level differ, so tuning the
+    dictionary for the compression level can help quite a bit.
+    """
+
+    # Check arguments' type
+    if not isinstance(zstd_dict, ZstdDict):
+        raise TypeError('zstd_dict argument should be a ZstdDict object.')
+    if not isinstance(dict_size, int):
+        raise TypeError('dict_size argument should be an int object.')
+    if not isinstance(level, int):
+        raise TypeError('level argument should be an int object.')
+
+    samples = tuple(samples)
+    chunks = b''.join(samples)
+    chunk_sizes = tuple(_nbytes(sample) for sample in samples)
+    if not chunks:
+        raise ValueError("The samples are empty content, can't finalize"
+                         "dictionary.")
+    dict_content = _finalize_dict(zstd_dict.dict_content,
+                                  chunks, chunk_sizes,
+                                  dict_size, level)
+    return ZstdDict(dict_content)
+
+def compress(data, level=None, options=None, zstd_dict=None):
+    """Return Zstandard compressed *data* as bytes.
+
+    *level* is an int specifying the compression level to use, defaulting to
+    COMPRESSION_LEVEL_DEFAULT
+
+    *options* is a dict object that contains advanced compression
+    parameters. See CParameter for more on options.
+
+    *zstd_dict* is a ZstdDict object, a pre-trained Zstandard dictionary. See
+    the function train_dict for how to train a ZstdDict on sample data.
+
+    For incremental compression, use a ZstdCompressor instead.
+    """
+    comp = ZstdCompressor(level=level, options=options, zstd_dict=zstd_dict)
+    return comp.compress(data, mode=ZstdCompressor.FLUSH_FRAME)
+
+def decompress(data, zstd_dict=None, options=None):
+    """Decompress one or more frames of Zstandard compressed *data*.
+
+    *zstd_dict* is a ZstdDict object, a pre-trained Zstandard dictionary. See
+    the function train_dict for how to train a ZstdDict on sample data.
+
+    *options* is a dict object that contains advanced compression
+    parameters. See DParameter for more on options.
+
+    For incremental decompression, use a ZstdDecompressor instead.
+    """
+    results = []
+    while True:
+        decomp = ZstdDecompressor(options=options, zstd_dict=zstd_dict)
+        results.append(decomp.decompress(data))
+        if not decomp.eof:
+            raise ZstdError("Compressed data ended before the "
+                            "end-of-stream marker was reached")
+        data = decomp.unused_data
+        if not data:
+            break
+    return b"".join(results)
+
+
+class CParameter(enum.IntEnum):
+    """Compression parameters"""
```

Comment:
bikeshed: perhaps `CompressionOptions` and `DecompressionOptions`? These are only used as keys to the `options` dict, as far as I can tell.

It would be good to expand `C` to `Compress(ion)?` though (regardless of Parameter/Options/etc) if we  it makes it more obvious what this is for. When first skimming this I assumed it had to do with the C language, rather than C for Compression. Same applies for `D`.

Reply:
I guess my main concern is that `CompressionOptions.compression_level` seems rather long to type. One thing I discussed with @Rogdham was setting CParameter/DParameter members as module attributes. Past discussion here: https://github.com/emmatyping/cpython/issues/5

Reply:
> It would be good to expand C to Compress(ion)?

I am sightly in favor of expending the name, to avoid any confusion.

> Explicit is better than implicit.
> ‒ The Zen of Python, by Tim Peters

People finding it too long in their code can always do
```python
from compression.zstd import CompressionParameter as CParam
```

---

> Parameter/Options/etc

I would be in favor of “Parameter” here.

In the [zstd manual](https://facebook.github.io/zstd/zstd_manual.html), although they are naming the thing `ZSTD_cParameter`, they are speaking about “compression parameters” when using plain English.

Also personally, I think of “options” as the whole dict whereas “parameter” is a key of the dict. I am probably biased on this however.

Reply:
`CompressionParameter` and `DecompressionParameter` are fine with me. If I understand, these are advanced options anyway, so making them a little more verbose isn't the worst thing.

Reply:
Yeah I agree, having them be verbose is fine since they are more advanced.